### PR TITLE
Upgrade to `go1.20.1`

### PR DESCRIPTION
**Pull request by frouioui**
## Description

This PR upgrades the golang version used in vitess `main` from `go1.19.4` to `go1.20.1`. By doing this, we are fixing several vulnerabilities that are affecting our codebase.

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtctl/grpcvtctldserver/server_test.go [hunk 1]
```diff
@@ -11367,16 +11367,17 @@ func TestValidateVersionShard(t *testing.T) {
...
 
 	for _, tt := range tests {
+		curT := tt
 		t.Run(tt.name, func(t *testing.T) {
-			tt.setup()
-			resp, err := vtctld.ValidateVersionShard(ctx, tt.req)
-			if tt.shouldErr {
+			curT.setup()
+			resp, err := vtctld.ValidateVersionShard(ctx, curT.req)
+			if curT.shouldErr {
```
Comment (deepthi) [inline go/vt/vtctl/grpcvtctldserver/server_test.go:11374]: @ajm188 main and release-16.0 have diverged. release-16.0 still has `t.Parallel()`. Is this change really required now that we no longer `t.Parallel()` in this test?

File: go/vt/vtorc/logic/topology_recovery.go [hunk 2]
```diff
@@ -254,7 +254,7 @@ func recoverDeadPrimary(ctx context.Context, analysisEntry inst.ReplicationAnaly
 	}
 
 	// Read the tablet information from the database to find the shard and keyspace of the tablet
-	tablet, err := inst.ReadTablet(analysisEntry.AnalyzedInstanceKey)
+	tablet, _ := inst.ReadTablet(analysisEntry.AnalyzedInstanceKey)
```
Comment (deepthi) [inline go/vt/vtorc/logic/topology_recovery.go:257]: @GuptaManan100 it seems wrong to ignore the error here. Should be exist the func altogether if we aren't able to read the tablet?

**Comment by deepthi** [comment]
Related: #12398 
I picked a couple of commits from that PR to try and get CI passing on this.

**Comment by deepthi** [comment]
Don't understand why Unit Test (mysql57) is failing with exit code 143. Merged in `main` to skip launchable so that we get all the output in CI.

File: go/vt/vtctl/grpcvtctldserver/server_test.go (hunk 1 above)
Comment (frouioui) [reply go/vt/vtctl/grpcvtctldserver/server_test.go]: I removed this change in 425362d26b and `golangci-lint` did not complain locally. Closing this for now.

**Comment by frouioui** [comment]
> Don't understand why Unit Test (mysql57) is failing with exit code 143. Merged in `main` to skip launchable so that we get all the output in CI.

Recently saw the same issue. [cedcaa3](https://github.com/vitessio/vitess/pull/12399/commits/cedcaa322982025d204c8e048f3397522a69a138) had all unit test workflows green.

File: go/vt/vtorc/logic/topology_recovery.go (hunk 2 above)
Comment (GuptaManan100) [reply go/vt/vtorc/logic/topology_recovery.go:257]: Yes, we should exit the function altogether. We can't proceed without the tablet's information. We'll get a panic when we try to read the keyspace and shard from this tablet information reference. If it was ignored before, that's a bug.

Comment (GuptaManan100) [reply go/vt/vtorc/logic/topology_recovery.go:257]:
```golang
if err != nil {
    return false, nil, err
}
```
is all we need after that line.

Comment (frouioui) [reply go/vt/vtorc/logic/topology_recovery.go:257]: Thank you @GuptaManan100, I'll port this change on `release-16.0`

Comment (frouioui) [reply go/vt/vtorc/logic/topology_recovery.go]: The change made in 42eb6db will fix https://github.com/vitessio/vitess/issues/12509

**Review by dbussink — APPROVED** [review]

**Review by shlomi-noach — APPROVED** [review]

File: go/bucketpool/bucketpool_test.go [hunk 3]
```diff
@@ -167,10 +167,12 @@ func TestPoolWeirdMaxSize(t *testing.T) {
 }
 
 func TestFuzz(t *testing.T) {
-	t.Skip()
 	maxTestSize := 16384
 	for i := 0; i < 20000; i++ {
 		minSize := rand.Intn(maxTestSize)
+		if minSize == 0 {
+			minSize = 1
```
Comment (deepthi) [inline go/bucketpool/bucketpool_test.go:174]: nit: I would have just skipped this iteration (`continue`) when minSize == 0.

**Review by deepthi — APPROVED** [review]
💯
